ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_parser.rs

```diff
@@ -0,0 +1,217 @@
+                num_transactions_to_fetch,
+            });
+            starting_version += num_transactions_to_fetch as u64;
+            num_fetches += 1;
+        }
+        batches
+    }
+
+    async fn fetch_raw_txns_with_retries(
+        context: Arc<Context>,
+        ledger_version: u64,
+        batch: TransactionBatchInfo,
+    ) -> Vec<TransactionOnChainData> {
+        loop {
+            match context.get_transactions(
+                batch.start_version,
+                batch.num_transactions_to_fetch,
+                ledger_version,
+            ) {
+                Ok(raw_txns) => return raw_txns,
+                Err(_err) => {
+                    continue;
+                },
+            }
+        }
+    }
+
+    /// Parses the table information from the raw transactions before converting to the api transactions,
+    /// optionally backup the rocksdb to gcs depending on epoch advancement or not.
+    async fn parse_table_info(
+        context: Arc<Context>,
+        raw_txns: Vec<TransactionOnChainData>,
+        db_writer: Arc<dyn DbWriter>,
+    ) -> Result<(), Error> {
+        if raw_txns.is_empty() {
+            return Ok(());
+        }
+
+        let start_millis = chrono::Utc::now().naive_utc();
+        let first_version = raw_txns.first().map(|txn| txn.version).unwrap();
+        let ledger_chain_id = context.chain_id().id();
+        let (_, _, block_event) = context
+            .db
+            .get_block_info_by_version(first_version)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Could not get block_info for start version {}",
+                    first_version,
+                )
+            });
+        let block_event_epoch = block_event.epoch();
+
+        let write_sets: Vec<WriteSet> = raw_txns.iter().map(|txn| txn.changes.clone()).collect();
+        let write_sets_slice: Vec<&WriteSet> = write_sets.iter().collect();
+        db_writer
+            .clone()
+            .index(
+                context.db.clone(),
+                first_version,
+                &write_sets_slice,
+                block_event_epoch,
+            )
+            .expect("Failed to process write sets and index to the table info rocksdb");
+        let fetch_millis = (chrono::Utc::now().naive_utc() - start_millis).num_milliseconds();
+
+        info!(
+            table_info_first_version = first_version,
+            current_epoch = block_event_epoch,
+            ledger_chain_id = ledger_chain_id,
+            write_sets_bcs_size = bcs::serialized_size(&write_sets).unwrap(),
+            table_info_parsing_millis_per_batch = fetch_millis,
+            num_transactions = raw_txns.len(),
+        );
+
+        Ok(())
+    }
+
+    pub fn set_highest_known_version(&mut self) -> anyhow::Result<()> {
+        let info = self.context.get_latest_ledger_info_wrapped()?;
+        self.highest_known_version = info.ledger_version.0;
+        Ok(())
```

Comment:
Why do you need this? The storage db should have the latest version I don't think that you need to track a separate variable in memory. Unless performance? Anyway I wouldn't do it unless there's a good reason to.

Reply:
multiple places use `self.highest_known_version`, it won't be efficient to query against ledgerinfo everytime we need it.

Reply:
why does multiple place query it? you should only need to query it once per loop. And yes it's a storage pull so it should be slightly less efficient than in memory, but it adds unnecessary complexity and potentially cause bugs.

Reply:
Reading the code, I still maintain that this is unnecessary. There are only 2 places that actually use this, the first one at the beginning of the loop to decide when to batch and the second one at the end. I don't even think that the second one is needed.